[SPARK-10548] [SPARK-10563] [SQL] Fix concurrent SQL executions

andrewor14 opened the pull request:
_Note: this is for master branch only._ The fix for branch-1.5 is at #8721.

The query execution ID is currently passed from a thread to its children, which is not the intended behavior. This led to `IllegalArgumentException: spark.sql.execution.id is already set` when running queries in parallel, e.g.:

```
(1 to 100).par.foreach { _ =>
  sc.parallelize(1 to 5).map { i => (i, i) }.toDF("a", "b").count()
}
```

The cause is `SparkContext`'s local properties are inherited by default. This patch adds a way to exclude keys we don't want to be inherited, and makes SQL go through that code path.

Commit:
Exclude certain local properties from being inherited

such as, cough cough, the SQL execution ID. This was a problem
because scala's parallel collections spawns threads as children
of the existing threads, causing the execution ID to be inherited
when it shouldn't be.

Commit:
Fix remove from Properties + add tests

Because java.util.Properties' remove method takes in an Any
instead of a String, there were some issues with matching the
key's hashCode, so removing was not successful in unit tests.

Instead, this commit fixes it by manually filtering out the keys
and adding them to the child thread's properties.

andrewor14 commented:
@davies @zsxwing FYI

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42307/consoleFull) for   PR 8710 at commit [`3ec715c`](https://github.com/apache/spark/commit/3ec715c4e5af5fa8d5e58c7aa93d01cd09970ae7).

SparkQA commented:
[Test build #42307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42307/console) for   PR 8710 at commit [`3ec715c`](https://github.com/apache/spark/commit/3ec715c4e5af5fa8d5e58c7aa93d01cd09970ae7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42307/
Test FAILed.

Commit:
Fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented on sql/core/src/test/scala/org/apache/spark/sql/execution/SQLExecutionSuite.scala:
```diff
@@ -0,0 +1,55 @@
+      // We need to run this multiple times to ensure new threads are spawned. Without the fix
+      // for SPARK-10548, this usually fails on the second try.
+      val df = sparkContext.parallelize(1 to 5).map { i => (i, i) }.toDF("a", "b")
+      (1 to 10).par.foreach { _ => df.count() }
```
I thought DataFrame is not thread-safe and should not be used like this.

marmbrus replied:
What makes you think that?  SparkContext/SQLContext/DataFrames should be threadsafe.

zsxwing replied:
I saw your reply here: http://mail-archives.us.apache.org/mod_mbox/spark-user/201504.mbox/%3CCAAswR-5m2EGZZU0_Lns2=tT9k+ydR4AJ9c1TNb7Cs31nbDzsEw@mail.gmail.com%3E

marmbrus replied:
Ah, I think we have fixed that, and if not I would also consider that a bug :)

SparkQA commented:
[Test build #42308 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42308/consoleFull) for   PR 8710 at commit [`d48c114`](https://github.com/apache/spark/commit/d48c11400b11e58d6ee3a4f5d9c330dd1736f25c).

zsxwing commented:
> The cause is SparkContext's local properties are inherited by default.

I just realized this is not the cause. There is no new thread between `sc.setLocalProperty(EXECUTION_ID_KEY, executionId.toString)` and `sc.setLocalProperty(EXECUTION_ID_KEY, null)` in your example. Actually, I found `sc.setLocalProperty(EXECUTION_ID_KEY, null)` sometimes cannot clear `EXECUTION_ID_KEY`. I will investigate it.

zsxwing commented:
Ah, I see. So the issue is:

1. Thread A creates a new Thread B
2. Thread A starts to run a query
3. Thread A is running a query
4. Thread B saw the execution id in Thread A's properties, then it will throw an exception.

zsxwing commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -348,10 +348,27 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
 
   // Thread Local variable that can be used by users to pass information down the stack
   private val localProperties = new InheritableThreadLocal[Properties] {
-    override protected def childValue(parent: Properties): Properties = new Properties(parent)
```
@andrewor14 I'm thinking maybe we should not use `new Properties(parent)` here. Instead, always copy the parent's Properties to the child's Properties. Do you think if the child thread needs to see the further changes to the parent thread's Properties after creating?

This is really confusing when using `Executor` like `ForkJoinPool`, in which thread A creates thread B but thread B is not a child of thread A. But thread B still can see the changes in thread A.

/cc @jerryshao since you added this line.

jerryshao replied:
Hi @zsxwing , reasons to use `InheritableThreadLocal` can be seen here (https://github.com/mesos/spark/pull/937). mainly it is used for Spark Streaming with FIFO scheduling strategy.

zsxwing replied:
I see. However, `new Properties(parent)` keeps a reference to parent rather than copying them. So If we make any change to the parent thread's properties after creating the child thread, the child thread will see them.

Is it necessary that the child thread keeps track of the following updates of the parent thread's properties? I think copying them would be more reasonable.

jerryshao replied:
Yes I agree with you that copying is more reasonable, for now I cannot image any scenario which requires to keep track of parent's properties, I think it is OK for me to change it, we can always fix this if there's any special scenario.

zsxwing replied:
@jerryshao Thanks :)

@andrewor14 how about just copying the parent properties rather than adding `nonInheritedLocalProperties`? It looks simpler.

andrewor14 replied:
I agree, I was actually going to do it in a separate patch. Incidentally @tdas @JoshRosen and I just talked about this last night and we all agreed to make it do a clone instead so the semantics are simpler.

However, my one concern is that doing so will change semantics for non-SQL users in 1.5.1, so my proposal is the following: I will make the changes in this patch and merge this patch ONLY into master. Then I'll create a new patch for branch 1.5 that will have the current changes (the ones where we don't clone except for SQL). I think that's the safest way forward.

andrewor14 replied:
OK, I have updated this in the latest commit

SparkQA commented:
[Test build #42308 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42308/console) for   PR 8710 at commit [`d48c114`](https://github.com/apache/spark/commit/d48c11400b11e58d6ee3a4f5d9c330dd1736f25c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ExecutorLostFailure(execId: String, isNormalExit: Boolean = false)`
  * `class ExecutorLossReason(val message: String) extends Serializable `
  * `case class ExecutorExited(exitCode: Int, isNormalExit: Boolean, reason: String)`
  * `  case class RemoveExecutor(executorId: String, reason: ExecutorLossReason)`
  * `  case class GetExecutorLossReason(executorId: String) extends CoarseGrainedClusterMessage`
  * `case class ConvertToSafeNode(conf: SQLConf, child: LocalNode) extends UnaryLocalNode(conf) `
  * `case class ConvertToUnsafeNode(conf: SQLConf, child: LocalNode) extends UnaryLocalNode(conf) `
  * `case class FilterNode(conf: SQLConf, condition: Expression, child: LocalNode)`
  * `case class HashJoinNode(`
  * `case class LimitNode(conf: SQLConf, limit: Int, child: LocalNode) extends UnaryLocalNode(conf) `
  * `abstract class LocalNode(conf: SQLConf) extends TreeNode[LocalNode] with Logging `
  * `abstract class LeafLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `abstract class UnaryLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `abstract class BinaryLocalNode(conf: SQLConf) extends LocalNode(conf) `
  * `case class ProjectNode(conf: SQLConf, projectList: Seq[NamedExpression], child: LocalNode)`
  * `case class SeqScanNode(conf: SQLConf, output: Seq[Attribute], data: Seq[InternalRow])`
  * `case class UnionNode(conf: SQLConf, children: Seq[LocalNode]) extends LocalNode(conf) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42308/
Test FAILed.

andrewor14 commented:
> Ah, I see. So the issue is:
>
> Thread A creates a new Thread B
> Thread A starts to run a query (set the execution id property)
> Thread A is running a query
> Thread B sees the execution id in Thread A's properties set by step 2, then it will throw an exception.

correct!

Commit:
Merge branch 'master' of github.com:apache/spark into concurrent-sql-executions

Commit:
Always clone parent properties

... to make the behavior more consistent in SQL vs non-SQL cases.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
As of the latest commit this patch should only be merged into master. I consider the fix for SPARK-10563 a little too risky for 1.5.1, so I will open a separate patch for branch-1.5 without that fix.

SparkQA commented:
[Test build #42342 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42342/consoleFull) for   PR 8710 at commit [`5297f79`](https://github.com/apache/spark/commit/5297f795e4e0ad5cf3bb6ff520312c2b82684f86).

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42357/consoleFull) for   PR 8710 at commit [`5297f79`](https://github.com/apache/spark/commit/5297f795e4e0ad5cf3bb6ff520312c2b82684f86).

SparkQA commented:
[Test build #42342 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42342/console) for   PR 8710 at commit [`5297f79`](https://github.com/apache/spark/commit/5297f795e4e0ad5cf3bb6ff520312c2b82684f86).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class MultilayerPerceptronClassificationModel(JavaModel):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol):`
  * `class MinMaxScalerModel(JavaModel):`
  * `        ("thresholds", "Thresholds in multi-class classification to adjust the probability of " +`
  * `class HasElasticNetParam(Params):`
  * `class HasFitIntercept(Params):`
  * `class HasStandardization(Params):`
  * `class HasThresholds(Params):`
  * `    thresholds = Param(Params._dummy(), "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `        self.thresholds = Param(self, "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42342/
Test PASSed.

SparkQA commented:
[Test build #42357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42357/console) for   PR 8710 at commit [`5297f79`](https://github.com/apache/spark/commit/5297f795e4e0ad5cf3bb6ff520312c2b82684f86).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42357/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into concurrent-sql-executions

Conflicts:
	core/src/test/scala/org/apache/spark/ThreadingSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42361/consoleFull) for   PR 8710 at commit [`3c00cc6`](https://github.com/apache/spark/commit/3c00cc64938d365abb783bb4041f6fdeff48d83f).

SparkQA commented:
[Test build #42361 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42361/console) for   PR 8710 at commit [`3c00cc6`](https://github.com/apache/spark/commit/3c00cc64938d365abb783bb4041f6fdeff48d83f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class MultilayerPerceptronClassificationModel(JavaModel):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol):`
  * `class MinMaxScalerModel(JavaModel):`
  * `        ("thresholds", "Thresholds in multi-class classification to adjust the probability of " +`
  * `class HasElasticNetParam(Params):`
  * `class HasFitIntercept(Params):`
  * `class HasStandardization(Params):`
  * `class HasThresholds(Params):`
  * `    thresholds = Param(Params._dummy(), "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `        self.thresholds = Param(self, "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `case class IntersectNode(conf: SQLConf, left: LocalNode, right: LocalNode)`
  * `case class SampleNode(`
  * `case class TakeOrderedAndProjectNode(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42361/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -348,10 +348,24 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+  /**
+   * Keys of local properties that should not be inherited by children threads.
+   */
+  private[spark] val nonInheritedLocalProperties: HashSet[String] = new HashSet[String]
```
Exposing a mutable `HashSet` in the thread-safe `SparkContext` looks dangerous. Actually, I suggest we don't need to add `nonInheritedLocalProperties` in the master branch. How about just cloning the parent properties without adding the `nonInheritedLocalProperties` logic?

andrewor14 replied:
the whole point of SPARK-10548 is to avoid inheriting the SQL execution ID. How can we avoid that with just cloning?

andrewor14 replied:
I made this private in the latest commit and added a setter method for it. Does this address your concern?

Commit:
Merge branch 'master' of github.com:apache/spark into concurrent-sql-executions

Commit:
Limit scope of mutable set

AmplabJenkins commented:
Merged build triggered.

andrewor14 commented:
@zsxwing I just noticed a potential source of confusion. If I understand correctly your view is that we should just clone the properties instead of having the `nonInheritedLocalProperties`. However, just cloning the properties won't fix SPARK-10548, because the issue is that the execution ID is passed directly to the child thread, NOT that it is mutated after the child thread is spawned.

Does that make sense? Please let me know if I'm missing something.

zsxwing commented:
> the execution ID is passed directly to the child thread

I believe that this is not the cause of `SPARK-10548`. The cause of `SPARK-10548` is the child thread can see the execution id that is set by the parent thread after the child thread is spawned.

If there is no execution id in the local properties when creating a child thread and we change it to clone the properties, then the child thread won't see the execution id that is set by the parent thread.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42379/consoleFull) for   PR 8710 at commit [`35bb6f0`](https://github.com/apache/spark/commit/35bb6f08befdab26edf5535f27f0b6a405f142f2).

SparkQA commented:
[Test build #42379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42379/console) for   PR 8710 at commit [`35bb6f0`](https://github.com/apache/spark/commit/35bb6f08befdab26edf5535f27f0b6a405f142f2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42379/
Test PASSed.

Commit:
Simplify fix by cloning properties on inherit

The fix for SPARK-10548 can be simplified by just cloning the
parent properties on inherit rather than excluding specific
properties from ever being inherited. This is safe because the
child thread must be created BEFORE the parent thread runs a
query.

Commit:
Merge branch 'master' of github.com:apache/spark into concurrent-sql-executions

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
> If there is no execution id in the local properties when creating a child thread and we change it to clone the properties, then the child thread won't see the execution id that is set by the parent thread.

Ah, I see. You're saying the child thread must be created *before* the query is run, not while it's running. That makes sense. Previously I accounted for the case where the child thread is created in the middle of the query, which I suppose is not possible. I have updated the code based on your suggestion.

SparkQA commented:
[Test build #42441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42441/consoleFull) for   PR 8710 at commit [`b4bcc3c`](https://github.com/apache/spark/commit/b4bcc3c76b0cea33cbc38997f160f1e128d66a45).

andrewor14 commented:
@zsxwing Alright I have updated it. Please have another look. I also updated the one for branch-1.5 (#8721), which has all of the changes here except the new behavior is triggered only in SQL.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42447 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42447/consoleFull) for   PR 8710 at commit [`fce3819`](https://github.com/apache/spark/commit/fce38190f1252ddc8ac20777d121a290a0155234).

Commit:
Merge branch 'master' of github.com:apache/spark into concurrent-sql-executions

Conflicts:
	core/src/test/scala/org/apache/spark/ThreadingSuite.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42452 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42452/consoleFull) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).

SparkQA commented:
[Test build #42441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42441/console) for   PR 8710 at commit [`b4bcc3c`](https://github.com/apache/spark/commit/b4bcc3c76b0cea33cbc38997f160f1e128d66a45).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42441/
Test FAILed.

SparkQA commented:
[Test build #42447 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42447/console) for   PR 8710 at commit [`fce3819`](https://github.com/apache/spark/commit/fce38190f1252ddc8ac20777d121a290a0155234).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultilayerPerceptronClassifier(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class MultilayerPerceptronClassificationModel(JavaModel):`
  * `class MinMaxScaler(JavaEstimator, HasInputCol, HasOutputCol):`
  * `class MinMaxScalerModel(JavaModel):`
  * `        ("thresholds", "Thresholds in multi-class classification to adjust the probability of " +`
  * `class HasElasticNetParam(Params):`
  * `class HasFitIntercept(Params):`
  * `class HasStandardization(Params):`
  * `class HasThresholds(Params):`
  * `    thresholds = Param(Params._dummy(), "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `        self.thresholds = Param(self, "thresholds", "Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values >= 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class' threshold.")`
  * `case class Stddev(child: Expression) extends StddevAgg(child) `
  * `case class StddevPop(child: Expression) extends StddevAgg(child) `
  * `case class StddevSamp(child: Expression) extends StddevAgg(child) `
  * `abstract class StddevAgg(child: Expression) extends AlgebraicAggregate `
  * `abstract class StddevAgg1(child: Expression) extends UnaryExpression with PartialAggregate1 `
  * `case class Stddev(child: Expression) extends StddevAgg1(child) `
  * `case class StddevPop(child: Expression) extends StddevAgg1(child) `
  * `case class StddevSamp(child: Expression) extends StddevAgg1(child) `
  * `case class ComputePartialStd(child: Expression) extends UnaryExpression with AggregateExpression1 `
  * `case class ComputePartialStdFunction (`
  * `case class MergePartialStd(`
  * `case class MergePartialStdFunction(`
  * `case class StddevFunction(`
  * `case class IntersectNode(conf: SQLConf, left: LocalNode, right: LocalNode)`
  * `case class SampleNode(`
  * `case class TakeOrderedAndProjectNode(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42447/
Test FAILed.

SparkQA commented:
[Test build #1751 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1751/consoleFull) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).

SparkQA commented:
[Test build #1753 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1753/consoleFull) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).

SparkQA commented:
[Test build #1752 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1752/consoleFull) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).

SparkQA commented:
[Test build #42452 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42452/console) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42452/
Test FAILed.

SparkQA commented:
[Test build #1753 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1753/console) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1751 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1751/console) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1752 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1752/console) for   PR 8710 at commit [`75a8d90`](https://github.com/apache/spark/commit/75a8d904f9935226df6872d15f99c8344615213c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

zsxwing commented:
LGTM

andrewor14 commented:
Thanks, I'm merging this into master.

d-ee commented:
This still seems to be around.
We're using Spark 1.5.2.

java.lang.IllegalArgumentException: "spark.sql.execution.id is already set"

Trace:
org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:87)
org.apache.spark.sql.DataFrame.withNewExecutionId(DataFrame.scala:1903)
org.apache.spark.sql.DataFrame.collect(DataFrame.scala:1384)
org.apache.spark.sql.DataFrame.head(DataFrame.scala:1314)
org.apache.spark.sql.DataFrame.head(DataFrame.scala:1321)
org.apache.spark.sql.DataFrame.first(DataFrame.scala:1328)

zsxwing commented:
@d-ee do you have a reproducer? Let's move the discussion to JIRA instead of here.

ljwagerfield commented:
We're seeing this exception too. We're also running our operations in serial (at least on the surface it seems as if we are). If we execute a `df.save` operation in a `Future` and wait for that `Future` to complete, then all `df.save` operations we perform within subsequent `Future`s will fail.

This specifically happens when we load Avro files from S3 and save them as Parquet back to S3. The loading works fine but the saving fails on 2nd attempt. Furthermore, if we simply generate a `DataFrame` from an in-memory list (so we're not loading from S3 - only saving to S3) then the error goes away... I'm not sure how helpful this is.

We're using Java 1.8, Scala 2.10.5, with our Spark codebase at commit https://github.com/apache/spark/commit/15de51c238a7340fa81cb0b80d029a05d97bfc5c.

Our exact reproduction steps are:

**1. Run a Spark Shell with appropriate dependencies**
```
./spark-shell --packages com.amazonaws:aws-java-sdk:1.10.75,org.apache.hadoop:hadoop-aws:2.7.2,com.databricks:spark-avro_2.10:2.0.1
```

**2. Run the following setup code within the shell**
```
import scala.concurrent.{ExecutionContext, Future}
import scala.concurrent.ExecutionContext.Implicits.global
import sqlContext.implicits._
import org.apache.spark.sql._
implicit val sqlContext = new org.apache.spark.sql.SQLContext(sc)

val hadoopConf = sc.hadoopConfiguration;
hadoopConf.set("fs.s3.impl", "org.apache.hadoop.fs.s3native.NativeS3FileSystem")
hadoopConf.set("fs.s3.awsAccessKeyId", "...")
hadoopConf.set("fs.s3.awsSecretAccessKey", "...")

val df = sqlContext.read.format("com.databricks.spark.avro").load("s3://bucket/input.avro")

def doWrite() {
    df.write.format("org.apache.spark.sql.parquet").mode(SaveMode.Overwrite).save("s3://bucket/output")
}
```

**3. Run this _twice_ - but leaving time for the first execution to finish (so the operations are serialised)**
```
Future { doWrite(); println("SUCCEEDED") }.recover { case e: Throwable => println("FAILED: " + e.getMessage()); e.printStackTrace() }
```

**Result:**
```
spark.sql.execution.id is already set
java.lang.IllegalArgumentException: spark.sql.execution.id is already set
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:87)
	at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelation.run(InsertIntoHadoopFsRelation.scala:108)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult$lzycompute(commands.scala:58)
	at org.apache.spark.sql.execution.ExecutedCommand.sideEffectResult(commands.scala:56)
	at org.apache.spark.sql.execution.ExecutedCommand.doExecute(commands.scala:70)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:132)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$5.apply(SparkPlan.scala:130)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:130)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:55)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:55)
	at org.apache.spark.sql.execution.datasources.ResolvedDataSource$.apply(ResolvedDataSource.scala:256)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:148)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:139)
	at $line38.$read$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.doWrite(<console>:41)
	at $line40.$read$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$anonfun$2.apply$mcV$sp(<console>:43)
	at $line40.$read$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$anonfun$2.apply(<console>:43)
	at $line40.$read$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$anonfun$2.apply(<console>:43)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

zsxwing commented:
@ljwagerfield it should be fixed in https://github.com/apache/spark/pull/11586

andrewor14 commented:
Yes, unfortunately that is only available in the upcoming 2.0 so you will have to upgrade to fix the problem.
